lxd/instances_put.go

```diff
@@ -0,0 +1,319 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/lxc/lxd/lxd/cgroup"
+	"github.com/lxc/lxd/lxd/db"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/operations"
+	"github.com/lxc/lxd/lxd/response"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/api"
+)
+
+func coalesceErrors(errors map[string]error) error {
+	if len(errors) == 0 {
+		return nil
+	}
+
+	errorMsg := "The following instances failed to update state:\n"
+	for instName, err := range errors {
+		errorMsg += fmt.Sprintf(" - Instance: %s: %v\n", instName, err)
+	}
+
+	return fmt.Errorf("%s", errorMsg)
+}
+
+// Update all instance states
+func containersPut(d *Daemon, r *http.Request) response.Response {
+	project := projectParam(r)
+
+	c, err := instance.LoadByProject(d.State(), project)
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	raw := api.InstancesPut{}
+	raw.Timeout = -1
+	if err := json.NewDecoder(r.Body).Decode(&raw); err != nil {
+		return response.BadRequest(err)
+	}
+
+	action := shared.InstanceAction(raw.Action)
+
+	var names []string
+	var instances []instance.Instance
+	for _, inst := range c {
+		switch action {
+		case shared.Freeze:
+			if !inst.IsRunning() {
+				continue
+			}
+		case shared.Restart:
+			if !inst.IsRunning() {
+				continue
+			}
+		case shared.Start:
+			if inst.IsRunning() {
+				continue
+			}
+		case shared.Stop:
+			if !inst.IsRunning() {
+				continue
+			}
+		case shared.Unfreeze:
+			if inst.IsRunning() {
+				continue
+			}
+		}
+
+		instances = append(instances, inst)
+		names = append(names, inst.Name())
+	}
+
+	// Don't mess with containers while in setup mode
+	<-d.readyChan
```

Comment:
@stgraber what is "setup" mode in this context (for my own info)?

Reply:
IIRC it's the time until all instances and tasks are running again.
During this time some API endpoints are responsive, others wait.

Reply:
@stgraber ah yes so it is. Wonder if it would be better to reject API requests if LXD server isn't ready rather than blocking here (presumably this would apply to the other API route we use this channel too).